syntax = "proto3";

package org.couchers.api.groups;

import "google/protobuf/empty.proto";
import "google/protobuf/timestamp.proto";

import "pb/discussions.proto";
import "pb/pages.proto";

service Groups {
  // This is a secure service: a user needs to be authenticated and not jailed to call functions here, refer to auth.proto and jail.proto

  rpc GetGroup(GetGroupReq) returns (Group) {
    // Get info about a group
  }

  rpc ListAdmins(ListAdminsReq) returns (ListAdminsRes) {
    // Gets a paginated list of group admins
  }

  rpc ListMembers(ListMembersReq) returns (ListMembersRes) {
    // Gets a paginated list of group members, including admins
  }

  rpc ListPlaces(ListPlacesReq) returns (ListPlacesRes) {
    // Returns a paginated list of places associated with a group
  }

  rpc ListGuides(ListGuidesReq) returns (ListGuidesRes) {
    // Returns a paginated list of guides associated with a group
  }

  rpc ListEvents(ListEventsReq) returns (ListEventsRes) {
    // Returns a paginated events of guides associated with a group
  }

  rpc ListDiscussions(ListDiscussionsReq) returns (ListDiscussionsRes) {
    // Returns a paginated list of discussions associated with a group
  }

  rpc JoinGroup(JoinGroupReq) returns (google.protobuf.Empty) {
    // Join a group
    //
    // Raises: FAILED_PRECONDITION if the user is already part of the group
  }

  rpc LeaveGroup(LeaveGroupReq) returns (google.protobuf.Empty) {
    // Leave a group
    //
    // Raises: FAILED_PRECONDITION if the user is not part of the group
  }
}

message CommunityParent {
  int64 community_id = 1;
  string name = 2;
  // a short URL fragment generated from the name, e.g. New York City: a guide to Phở might become new-york-city-guide-to-pho
  string slug = 3;
  string description = 4;
}

message GroupParent {
  int64 group_id = 1;
  string name = 2;
  // a short URL fragment generated from the name, e.g. New York City: a guide to Phở might become new-york-city-guide-to-pho
  string slug = 3;
  string description = 4;
}

message Parent {
  oneof parent {
    CommunityParent community = 1;
    GroupParent group = 2;
  }
}

message Group {
  int64 group_id = 1;
  string name = 2;
  // a short URL fragment generated from the name, e.g. New York City: a guide to Phở might become new-york-city-guide-to-pho
  string slug = 3;
  string description = 4;
  google.protobuf.Timestamp created = 5;
  // list of parents, ordered according to inclusion, e.g. World, Australia, (East Coast Hitchhikers)
  repeated Parent parents = 6;
  org.couchers.api.pages.Page main_page = 7;
  bool member = 8;
  bool admin = 9;
  uint32 member_count = 10;
  uint32 admin_count = 11;
}

message GetGroupReq {
  int64 group_id = 1;
}

message ListAdminsReq {
  int64 group_id = 1;

  uint32 page_size = 2;
  string page_token = 3;
}

message ListAdminsRes {
  repeated int64 admin_user_ids = 1;

  string next_page_token = 2;
}

message ListMembersReq {
  int64 group_id = 1;

  uint32 page_size = 2;
  string page_token = 3;
}

message ListMembersRes {
  repeated int64 member_user_ids = 1;

  string next_page_token = 2;
}

message ListPlacesReq {
  int64 group_id = 1;

  uint32 page_size = 2;
  string page_token = 3;
}

message ListPlacesRes {
  repeated org.couchers.api.pages.Page places = 1;

  string next_page_token = 2;
}

message ListGuidesReq {
  int64 group_id = 1;

  uint32 page_size = 2;
  string page_token = 3;
}

message ListGuidesRes {
  repeated org.couchers.api.pages.Page guides = 1;

  string next_page_token = 2;
}

message ListEventsReq {
  int64 group_id = 1;

  uint32 page_size = 2;
  string page_token = 3;
}

message Event {
  // TODO
}

message ListEventsRes {
  repeated Event events = 1;

  string next_page_token = 2;
}

message ListDiscussionsReq {
  int64 group_id = 1;

  uint32 page_size = 2;
  string page_token = 3;
}

message ListDiscussionsRes {
  repeated org.couchers.api.discussions.Discussion discussions = 1;

  string next_page_token = 2;
}

message JoinGroupReq {
  int64 group_id = 1;
}

message LeaveGroupReq {
  int64 group_id = 1;
}
